File: DSA/Data_Structures/B-Tree.c
#include <stdio.h>
#include <stdlib.h>

/*
 * B-Trees
 *
 * This data structures a pretty useful to optimize disc operation because they are cache friendly because, each node can store a 
 * pack of keys which are stored contiguous.
 *
 * Operation:
 *
 * - Searching:
 *   This one is pretty stright forward:
 *      -> Iterate over the values of a node compare the target with the keys
 *      -> If not found move to the node whose value are between keys[i] keys[i +1]
 *      -> Repeat the operation until found and return true (or value associated with the key in real use scenarios) else 
 *         return false
 * - Adding a key:
 *    # For the case the node is not overfull after insertion:
 *      -> Just put the key at the end of the keys at array and sort
 *    
 *    # When it is overfull after insertion
 * 
 *      -> Splint the node into two nodes, push the middle key to th parent node and if full again the repeat the process (Recursion). 
 *         If you arraive at the root node and it becomes full then:
 *
 *      -> Take the value at middle and make a new node whose children will be a node with values less than itself and another with
 *         the ones greater than itself. Only if we are at the root node!!!!!!
 *
 *
 *  - Removing a key:
 *      General first step search for the key and if found perform the deletion
 *
 *      Leaf cases
 *      # Case not under min after delete:
 *          -> nothing more 
 *      
 *      # Case underfull 1:
 *          -> take a key from one siebling the min for the right and the max for the left depending on the choice
 *             if possible -> they have enough keys. Take the key from the siebling and make it the new separator of the parent node 
 *             (the most right or left key) and then take the previous separator an use it to fill our node
 *      
 *      # Case underfull 2:
 *          -> Take the siebling, our node with its number of keys below the minimum, and the separator between of the parent node 
 *          and merge them into a node -> if the parent node now ends up with to few keys use repat the process of the cases 1 and 2 depending 
 *          on the sieblings (recursion)
 *
 *      Not leaf case:
 *      -> Take largest value on the left substree or the smallest on the right to become the new separator. But in the case of underfullness
 *      we already know what to do depending on the case. See above 
 * */
const int MAX_NUMBER_OF_KEYS = 4;
const int MAX_NUMBER_OF_CHILDREN = MAX_NUMBER_OF_KEYS + 1;
const int MIN_NUMBER_OF_KEYS = MAX_NUMBER_OF_KEYS / 2;

/*
 * This node object has an array of keys -> the values we want to store and use as delimiters and 
 * a array of pointers to nodes which represent the children.
 *
 * */
typedef struct Node {
    
   struct Node* childs[MAX_NUMBER_OF_CHILDREN];
   int keys[MAX_NUMBER_OF_KEYS];
   int count;

} Node;

typedef struct B_Tree {
    
    Node* root;
} B_Tree;

B_Tree* create_b_tree() {

    B_Tree* tree = (B_Tree*)malloc(sizeof(struct B_Tree));
    tree->root = NULL;

    return tree;
}

Node* create_node(int data) {
    
    Node* node = (Node*)malloc(sizeof(Node));
    
    node->count = 1;
    node->keys[0] = data;

    return NULL; 

}

int main() {
    printf("Verga");
    return 0;
}

File: DSA/Data_Structures/B-Tree.py

class B_Tree:
    
    def __init__(self, t) -> None:
        self.t = t 

    def search(self, u, key):
    
        i = 0 
        # linear scan to find the key
        while i < len(u.keys) and key > u.keys[i]: # as long as the key is greater
            i += 1

        if i < len(u.keys) and key == u.keys[i]: # if we are within the bound and the value of the keys are identical 
            return (u, i)

        if u.is_leaf:       # Not found
            return None 

        return self.search(u.children[i], key)

    def insertion(self, key):
        
        root = self.root 

        if root.is_full(self.t): # has 2t - 1 children 
            root = self.split_root()

        self.insert_not_full(root, key)

    def split_child(self, u, i):
        
        t = self.t 
        full_node = u.children[i] # pointer to ith child 

        new_node = Node()
        new_node.is_leaf = full_node.is_leaf
        new_node.keys = full_node.keys[t:] # copying keys after the median 

        if not full_node.is_leaf:
            new_node.children = full_node.children[t:] # same as above but for the pointers 

        u.children.insert(i + 1, new_node) # inserting the new node's key and value reference in the parent 
        u.keys.insert(i, full_node.keys[t - 1]) # inserting the median 

        full_node.keys = full_node.keys[:t - 1] # adjusting keys to go only until one before the median 
        full_node.children = full_node.children[:t] # same here but for the children 

    
    def split_root(self):
        
        new_root = Node()
        new_root.children = [self.root]
        
        self.root = new_root
        self.split_child(new_root, 0) # splitting the full root
        return new_root
    
    def insert_not_full(self, u, key):
        
        i = 0       
        while i < len(u.keys) and key > u.keys[i]: # linear scan
            i += 1

        if u.is_leaf:               # can insert
            u.keys.insert(i, key)

        else:
            if u.children[i].is_full(): # if the target node is full
                
                self.split_child(u, i)          # split the node
                i = i if key <= u.keys[i] else i + 1 # choosing between right and left for the new key 
            
            self.insert_not_full(u.children[i], key) # applying the algorithm again until the key can be inserted

        
    def deletion(self, key):
        pass
        
class Node:
    
    def __init__(self) -> None:
        self.is_leaf = False 
        self.keys = []
        self.children = []


    def is_full(self, t) -> bool:
        
        if len(self.keys) == 2*t - 1:
            return True 

        return False
